### PR TITLE
mt-generate: Be precise about final content after generation, v2

### PR DESCRIPTION
**Pull request by dexonsmith**
This is a completely reworked version of https://github.com/apple/apple-llvm-infrastructure-tools/pull/38, along with support for `start` directives and `tag` generation (although not pattern-matched yet).  See commit messages for details of the changes.

Note that the latter two pieces have not been well tested yet.

### CODE REVIEW

File: docs/mt-config.md [hunk 1]
```diff
@@ -82,14 +89,17 @@ This can be useful to generate consistent histories "as if" an automerger had
...
+            commits translated by `dir` directives.
 
-### `dir`: Generating commits from a split repo
+    ## # `dir`: Generating commits from a split repo
 
-```
-dir <branch> ( <name> | '-' ) <source>
+``` dir<branch>(<name> | '-')<source>
+                    dir<tag>(<name> | '-')<source>
+                        dir<tags - id>(<name> | '-')<source>/*
```
Comment (bogner) [inline docs/mt-config.md]: Looks like some strange formatting things happened here...

File: docs/mt-config.md [hunk 2]
```diff
@@ -118,11 +143,192 @@ The currently supported values for `<type>` are:
...
+ 7. Sort repeat commits, interleaving with other commits by commit timestamp.
+ 8. Determine whether to discard HEAD.
+    - Discard if there are no start commits.
+    - Discard if a start commit had to back up to merge base.
+    - Discard if repeat start commit was discarded.
+    - Discard if any content in HEAD is wrong vs. start commits.
+ 9. Discard repeat start commit if HEAD was discarded.
+10. Find a repeat start commit if necessary.
+    - List the newest first parent commit of the oldest repeat commit being
+      interleaved that modifies a repeated path.
```
Comment (bogner) [inline docs/mt-config.md]: "Discard repeat start commit" coming before "find a repeat start commit" is a little confusing. Would it make more sense to just have the "find if necessary" and clarify that it's not necessary if HEAD was discarded?

**Comment by dexonsmith** [comment]
> I mostly just skimmed the actual code changes, but overall it looks pretty reasonable. I am a bit confused over where to use repeat or where to just omit mentioning an upstream directory, but I think that'll become clear from the configs in the repo.

With the current tool I think you want a `repeat` as long as any of your upstreams has downstream history for a sub-project that you do not have downstream history for.

For example, consider the content in clang-tools-extra, which apple/master has downstream history for:

- If you're interleaving a downstream branch for clang-tools-extra then you'll have a `dir` for it.  Its content will be correct and fairly consistent, since your internal commits will be interleaved between clang and clang-tools-extra.  
- Else, if you use `repeat` on apple/master, the clang-tools-extra content will consistently match what's on apple/master's first-parent path, using generated merges interleaved with your downstream branch's changes to clang and llvm.
- Else (if you have no `repeat`), the content for clang-tools-extra will oscillate between matching llvm.org/master and apple/master.  Each merge commit that is translated will compare the SVN revision that the clang content is based on, and whichever one is higher will provide the clang-tools-extra content as well.

This oscillation would be weird.  I don't know how to do better without expensive operations on each translated commit.  Note that the above heuristic does work well for directories where neither apple/master nor any of its downstreams currently has a downstream branch (e.g., `llgo`), which is what informed its design.  (Also, this PR has not changed that heuristic at all.)

**Comment by dexonsmith** [comment]
@swift-ci please test

File: docs/mt-config.md (hunk 1 above)
Comment (dexonsmith) [reply docs/mt-config.md]: Yeah, that was me hitting my Vim key combination for `clang-format` as a reflex and not paying any attention.  Fixed it.

**Comment by dexonsmith** [comment]
@swift-ci test

**Comment by hyp** [comment]
@swift-ci test

**Pull request merged**